Incorporate distributions into inference

TimothyWillard opened the pull request:
### Describe your changes.

This pull request:

- Fixed bug in the call signature to `check_parameter_positivity` in `build_step_source_arg`.
- Fixed typos in call to `distribution_from_confuse_config` in `gempyor.inference_parameters`.
- Added `DistributionABC.support` abstract property to give the lower/upper bounds of a distribution.
- Incorporated `support` abstract property into the `InferenceParameters` class and modified the 'simple_usa_statelevel' example to demonstrate.
- Added `gempyor.NPI.helpers.SpatialGroups` to consistently represent modifier subpopulations and converted `get_spatial_groups` into the `from_subpopulations` class method for ease of creation from modifier configuration.

The remaining TODOs are:

- [x] Consolidate `hit_{l/u}bs` into `check_in_bound` (https://github.com/HopkinsIDD/flepiMoP/pull/601#discussion_r2291758148).
- [x] Remove the `{l/u}bs` attributes of the `InferenceParameters` class (https://github.com/HopkinsIDD/flepiMoP/pull/601#discussion_r2291749309).
- [x] Figure out how to set `_{lower/upper}` attributes from data for `FixedDistribution`, `UniformDistribution`, etc (https://github.com/HopkinsIDD/flepiMoP/pull/601#discussion_r2291749309).
- [x] Add back `{l/u}bs` attributes of the `InferenceParameters` class as private attributes (https://github.com/HopkinsIDD/flepiMoP/pull/601#discussion_r2294179574).
- [x] Add unit test for `InferenceParameters.inject_proposal` (https://github.com/HopkinsIDD/flepiMoP/pull/601#discussion_r2294186333).
- [x] Check if any documentation changes are required with respect to the changes in the `simple_usa_statelevel` example (https://github.com/HopkinsIDD/flepiMoP/pull/601#discussion_r2283075863).

### Does this pull request make any user interface changes? If so please describe.

The user interface changes are allowing the usage of distributions other than 'truncnorm' in the value for modifiers.

Those are reflected in updates to the documentation in ...

### What does your pull request address? Tag relevant issues.

This pull request addresses #598.

TimothyWillard committed:
Fix typo creating distributions in inference param

Fixed typo in the name of the distribution creation helper in two places
in `inference_parameter.py`. Also corrected call to use
`DistributionABC.sample` method because a specific size was required.

TimothyWillard committed:
Add `DistributionABC.support` property

Added a `support` property to the `DistributionABC` class which returns
a tuple of two numbers representing the upper and lower bounds
respectively. This property pulls said bounds from either the
`_lower/upper_bound` attributes if defined by the implementation or
`_lower/upper` properties if overridden by the implementation. Added to
the current examples to demonstrate usage. Minor modifications to the
commond distributions test was required.

TimothyWillard committed:
Incorporate bounds into inference parameters

Incorporated bounds provided by `DistributionABC.support` into the
`InferenceParameters` class for lower/upper bounds used in proposal
rejection. Also converted lower/upper bounds representation from list of
floats to numpy arrays of floats for performance reasons.

TimothyWillard committed:
Demonstrate inference with more than `truncnorm`

Used the 'simple_usa_statelevel' example to demonstrate how inference on
modifiers now supports more than 'truncnorm' distributions. Changed the
'truncnorm' usage in the seir modifiers section to use a 'gamma'
distribution with the same mean and variance.

TimothyWillard committed:
Provide subpop names to positive parameter check

Fix `TypeError` caused by providing the subpopulation populations,
`list[int]`, to `check_parameter_positivity` by providing the
subpopulation names, `list[str]`. This would cause an error when trying
to construct an error message when it tried to join the subpopulation
names together. The exception was: `sequence item 0: expected str
instance, numpy.int64 found`.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -72,6 +73,18 @@ def _sample_from_generator(
         """
         raise NotImplementedError
 
+    @property
+    @abstractmethod
+    def support(self) -> tuple[float | int, float | int]:
+        """
+        The theoretical support of the distribution as a (min, max) tuple.
+
+        Returns:
+            A tuple representing the upper and lower bounds in the distribution's
+            support.
+        """
+        raise NotImplementedError
+
```
I think we'd be better served here by adopting the wrapping delegate implementers - so have this return `(..., ...)` and then the subclasses indicate their lower / upper bounds.

is this something w/ can do with `ClassVar` definitions here, then values supplied in the sub classes - will pydantic enforce that developers define them in subclasses? ala, https://docs.pydantic.dev/latest/concepts/models/#class-variables

TimothyWillard replied:
I do not think that `ClassVar` is the right tool to reach for here since that will define the attribute for all instances of a class. For normal distributions this is fine, but for distributions with configurable support (i.e. truncated normal) this will not work. However, I can rework this PR to use the delegated implementers pattern that we've been using so far with `_lower` and `_upper` abstract properties.

pearsonca reviewed (commented):
I think this is probably fine, but I'd like to discuss repeating the pattern we've been using with the abstract class delegating to an internal implementing method (i.e. sample and _sample_impl)

seems like there are potentially natural opportunities with pydantic approaches

emprzy commented:
I like the added support to `distributions`. Will definitely keep this in mind as I'm working on #595 . Need to more thoroughly review in a bit

TimothyWillard committed:
Light refactor of `InferenceParameters` class

* Address suggestions from `pylint` to make the code more readable as
	well as add documentation/type hints.
* Added initial, highly targeted unit tests. Including for the
	`inject_proposal` method.
* Consolidated `hit_*` methods into the `check_in_bound` method.
* Made the `{l/u}bs` attributes private and extract them from the
	distribution object on behalf of the user in `add_single_parameter`.

pearsonca commented on examples/simple_usa_statelevel/simple_usa_statelevel.yml:
I'm generally fine with changing these, let's just make sure to check any connected documentation to ensure it still makes sense with the changes.

TimothyWillard replied:
No documentation updates are required to accommodate the example change. The 'simple_usa_statelevel' example is only mentioned in the HPC submission guide which just uses the config as an EMCEE inference example but does not discuss the details of the config contents.
```
(C flepiMoP/venv) ~/D/G/H/f/documentation (Light refactor of `InferenceParameters` class, wq, ae, model-output-api*) > rg -F 'simple_usa_statelevel'
gitbook/how-to-run/advanced-run-guides/running-on-a-hpc-with-slurm.md
99:$ export PROJECT_PATH="$FLEPI_PATH/examples/simple_usa_statelevel/"
115:    simple_usa_statelevel.yml
163:    simple_usa_statelevel.yml > simple_usa_statelevel_estimation.log 2>&1 & disown
166:In short, this command will submit 6 test jobs that will vary simulations and measure time and memory. The number of simulations will be used to project the required resources. The test jobs will range from 1/5 to 1/10 of the target job size. This command will take a bit to run because it needs to wait on these test jobs to finish running before it can do the analysis, so you can check on the progress by checking the output of the `simple_usa_statelevel_estimation.log` file.
186:    simple_usa_statelevel.yml
```

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -37,6 +38,12 @@ class DistributionABC(ABC, BaseModel):
+    _lower_bound: float | int | None = PrivateAttr(default=None)
+    _upper_bound: float | int | None = PrivateAttr(default=None)
+
+    def __call__(self) -> float | int:
+        """A shortcut for `self.sample(size=1)`."""
+        return self.sample(size=1).item()
```
Question: is this relocation is a linting based thing?

TimothyWillard replied:
This relocation was manual, I'm used to putting dunders near the top of the class. Happy to undo to clean up the diff if preferred.

pearsonca replied:
I'm fine with the relocation - also agree with this loc for dunders. Was asking more to see if this is something we should be adding to our style guide.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -97,6 +156,16 @@ def _sample_from_generator(
         """Sampling logic for fixed distributions."""
         return np.full(size, self.value)
 
+    @property
+    def _lower(self) -> float:
+        """The lower bound of the fixed distribution's support."""
+        return self.value
+
+    @property
+    def _upper(self) -> float:
+        """The upper bound of the fixed distribution's support."""
+        return self.value
```
for later: is there way we can use pydantic aliases for these kind of things?

pearsonca replied:
(applies to several potential options)

pearsonca replied:
cursory look at Alias indicates this almost entirely about serialization / deserialization, *not* object access.

we could potentially make lower / upper *the* fields, and then "value" (in this example) the alias.

but definitely this is a tabled item for future refinement - maybe incorporate in #603 ?

TimothyWillard replied:
Took another look at our options, I think we're best off waiting until https://github.com/pydantic/pydantic/pull/11685. I'll add a note to #603.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/inference_parameter.py:
```diff
@@ -51,29 +49,32 @@ def add_modifier(self, pname, ptype, parameter_config, subpops):
-                        ub=parameter_config["value"]["b"].get(float),
+                        pdist=dist,
+                        lb=lower,
+                        ub=upper,
```
if add_single_parameter is now always working on a pydantic distribution, which also now always comes with a support method, should we prune lb, ub arguments? Doesn't seem like we ever would want to pass these independently.

TimothyWillard replied:
I've pulled 87030fd9015b667d3abad24d9f0a1058a7c89439 from #608 to address.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/inference_parameter.py:
```diff
@@ -51,29 +49,32 @@ def add_modifier(self, pname, ptype, parameter_config, subpops):
-                        pdist=distribution_from_confuse_config(parameter_config["value"]),
-                        lb=parameter_config["value"]["a"].get(float),
-                        ub=parameter_config["value"]["b"].get(float),
+                        pdist=dist,
```
ibid

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/inference_parameter.py:
```diff
@@ -203,33 +203,46 @@ def draw_initial(self, n_draw=1):
-            bool: True if the proposal is within bounds, False otherwise.
+            `True` if the proposal is within bounds, `False` otherwise.
+        """
+        return np.logical_and(~self.hit_lbs(proposal), ~self.hit_ubs(proposal)).all()
```
are we using hit_lbs, hit_ubs elsewhere? really feels like this should be an element-wise lbs <= proposal & proposal <= ubs.

TimothyWillard replied:
No, I'll consolidate this down:
```
(C flepiMoP/venv) ~/D/G/H/flepiMoP (empty, oy, 2, incorporate-params-into-inference) > rg -F 'hit_lbs'
flepimop/gempyor_pkg/src/gempyor/inference_parameter.py
221:        return np.logical_and(~self.hit_lbs(proposal), ~self.hit_ubs(proposal)).all()
223:    def hit_lbs(self, proposal: npt.NDArray[np.float64]) -> npt.NDArray[np.bool_]:
(C flepiMoP/venv) ~/D/G/H/flepiMoP (empty, oy, 2, incorporate-params-into-inference) > rg -F 'hit_ubs'
flepimop/gempyor_pkg/src/gempyor/inference_parameter.py
221:        return np.logical_and(~self.hit_lbs(proposal), ~self.hit_ubs(proposal)).all()
235:    def hit_ubs(self, proposal: npt.NDArray[np.float64]) -> npt.NDArray[np.bool_]:
```

TimothyWillard replied:
I've pulled 87030fd9015b667d3abad24d9f0a1058a7c89439 from #608 to address.

pearsonca reviewed (commented):
A few minor items to consider - but nothing must-have

emprzy commented on flepimop/gempyor_pkg/src/gempyor/inference_parameter.py:
```diff
@@ -93,33 +94,32 @@ def add_modifier(self, pname, ptype, parameter_config, subpops):
                             ptype=ptype,
                             pname=pname,
                             subpop=sp,
-                            pdist=istribution_from_confuse_config(
```
typo, should read: `distribution_from_confuse_config(...`

emprzy replied:
Actually, I'm not seeing this issue reflected on the[ codebase view of the branch](https://github.com/HopkinsIDD/flepiMoP/blob/incorporate-params-into-inference/flepimop/gempyor_pkg/src/gempyor/inference_parameter.py). This is what i'm seeing in the PR:
<img width="605" height="32" alt="Screenshot 2025-08-22 at 11 29 07 AM" src="https://github.com/user-attachments/assets/2c3947fe-c7a4-4f2f-907b-3310529d4c63" />
So perhaps ignore? Not sure if this is just on my end...?

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/inference_parameter.py:
```diff
@@ -93,150 +157,112 @@ def add_modifier(self, pname, ptype, parameter_config, subpops):
-        boolean vector of True if the parameter is bigger than the upper bound and False if not
-        """
-        return np.array((proposal > self.ubs))
+        lower_bounds, upper_bounds = zip(*[dist.support for dist in self.pdists])
```
how often is check_in_bounds called? i'm not sure about the elimination of ubs, lbs as properties of this object - might still be useful to create and cache those with initialization / parsing, especially if this method is called many times post initialization.

pearsonca replied:
(still think they should go as arguments to the add_... method, but maybe do this extract => append there)

TimothyWillard replied:
Frequently. I'll make the adjustment.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/inference_parameter.py:
```diff
@@ -246,24 +272,16 @@ def inject_proposal(
-                    "value",
-                ] = proposal[p_idx]
+        for p_idx in range(len(self)):
+            df = snpi_df_mod if self.ptypes[p_idx] == "seir_modifiers" else hnpi_df_mod
```
love it - i assume this logic is working fine as pointer-/reference-like updating operation, rather than copy-on-assign to a subsequently ignored `df`?

pearsonca replied:
aside, i'm not clocking a proposal injection test

TimothyWillard replied:
Will add

pearsonca commented on flepimop/gempyor_pkg/tests/distributions/test_distributions_common.py:
```diff
@@ -1,16 +1,16 @@
-    def __call__(self) -> float | int:
-        """A shortcut for `self.sample(size=1)`."""
-        return self.sample(size=1).item()
+    _lower_bound: float = PrivateAttr(default=0.0)
```
where has `__call__` gone? should still want the mockobject to have that interface, i'd think?

TimothyWillard replied:
I removed it, the dummy object gets it by subclassing `DistributionABC`. No need to define `__call__` twice.

pearsonca replied:
ah, right - forget that got moved into parent class.

TimothyWillard committed:
Unit test and change `get_spatial_groups` API

Added unit tests and changed the API of
`gempyor.NPI.helpers.get_spatial_groups`. The API change moves away from
requiring a confuse configview to use the function, which pushes the
confuse wrangling to the caller. Improves unit testing and general
purpose usage and makes the function more stable as we move away from
confuse. Also simplified/improved the logic and enhanced the
documentation and type hints.

TimothyWillard committed:
Convert `SpatialGroups` to a `dataclass`

For ease of constructing and manipulating modifier spatial groups
converted the `SpatialGroups` typed dict to a frozen data class. This
allows for the continued typing support as well as adding support for
iterating over a `SpatialGroups` class in a consistent way. Also
migrated `get_spatial_groups` into the
`SpatialGroups.from_subpopulations` class method. Added corresponding
documentation and unit tests.
